service/resource/vaultcrt/update.go

```diff
@@ -25,10 +56,80 @@ func (r *Resource) NewUpdatePatch(ctx context.Context, obj, currentState, desire
 	return patch, nil
 }
 
-func (r *Resource) ApplyUpdateChange(ctx context.Context, obj, updateChange interface{}) error {
-	return nil
+func (r *Resource) newUpdateChange(ctx context.Context, obj, currentState, desiredState interface{}) (interface{}, error) {
+	customObject, err := key.ToCustomObject(obj)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+	currentSecret, err := toSecret(currentState)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+	desiredSecret, err := toSecret(desiredState)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	r.logger.Log("cluster", key.ClusterID(customObject), "debug", "finding out if the secret has to be updated")
+
+	var secretToUpdate *apiv1.Secret
+	{
+		TTL, err := time.ParseDuration(key.CrtTTL(customObject))
+		if err != nil {
+			return false, microerror.Mask(err)
+		}
+
+		renew, err := r.shouldCertBeRenewed(currentSecret, TTL, r.expirationThreshold)
+		if IsMissingAnnotation(err) {
+			// fall through
+		} else if err != nil {
+			return nil, microerror.Mask(err)
+		}
+
+		if renew {
+			secretToUpdate = desiredSecret
+
+			err := r.ensureVaultRole(customObject)
+			if err != nil {
+				return nil, microerror.Mask(err)
+			}
+
+			ca, crt, key, err := r.issueCertificate(customObject)
+			if err != nil {
+				return nil, microerror.Mask(err)
+			}
+
+			secretToUpdate.StringData[certificatetpr.CA.String()] = ca
+			secretToUpdate.StringData[certificatetpr.Crt.String()] = crt
+			secretToUpdate.StringData[certificatetpr.Key.String()] = key
+		}
+	}
+
+	r.logger.Log("cluster", key.ClusterID(customObject), "debug", "found out if the secret has to be updated")
+
+	return secretToUpdate, nil
 }
 
-func (r *Resource) newUpdateChange(ctx context.Context, obj, currentState, desiredState interface{}) (interface{}, error) {
-	return nil, nil
+func (r *Resource) shouldCertBeRenewed(secret *apiv1.Secret, TTL, threshold time.Duration) (bool, error) {
+	if secret == nil {
+		return false, microerror.Mask(missingAnnotationError)
+	}
+	if secret.Annotations == nil {
+		return false, microerror.Mask(missingAnnotationError)
+	}
+	a, ok := secret.Annotations[UpdateTimestampAnnotation]
+	if !ok {
+		return false, microerror.Mask(missingAnnotationError)
+	}
+
+	t, err := time.ParseInLocation(UpdateTimestampLayout, a, time.UTC)
+	if err != nil {
+		return false, microerror.Mask(err)
+	}
+
+	if t.Add(TTL).Add(-threshold).Before(r.currentTimeFactory()) {
```

Comment:
`t` is the time of the last update. `t` plus the certificate TTL is the point in time where the certificate expires. We want to renew certificates before they expire. Thus we substitute a threshold from the point in time of the expiration and compare this value against the current time.